### ISSUE TITLE
Custom locales causing issues with hreflang and other metas

### ISSUE THREAD

**Issue by steverowling** [khalwat/seomatic#290]
First off, I fully appreciate that this is an edge case!

I have a situation where I am using custom locales in Craft 2 to implement multi-site for one client. All the content is in English and I have duplicated the `en_gb` locale and extended it to create the custom locales to create three inked sites. The custom locales are in the format `en_gb_site1`, `en_gb_site2`, `en_gb_site3`.

All this works well within Craft for creating content for each site with sections targeting only one custom locale and sharing content between them with fully localised sections targeting all three custom locales.

The specific problems I'm facing are:

1. Since the site is localised, SEOmatic creates `hreflang="x-default"` tags for all pages. For pages in sections with only one locale enabled, I get this:

```
<link rel="canonical" href="http://site1.co.uk" />
<link rel="alternate" href="http://site1.co.uk/" hreflang="x-default" />
```

In this instance, I think the `hreflang-"x-default"` tag is redundant, since there is only one version of the page.

For pages in sections with all locales enabled, I get this:

```
<link rel="canonical" href="http://site1.co.uk/legal/privacy-policy" />
<link rel="alternate" href="http://site1.co.uk/legal/privacy-policy" hreflang="x-default" />
<link rel="alternate" hreflang="en-gb-site1" href="http://site1.co.uk/legal/privacy-policy" />
<link rel="alternate" hreflang="en-gb-site2" href="http://site2.co.uk/legal/privacy-policy" />
<link rel="alternate" hreflang="en-gb-site3" href="http://site3.co.uk/legal/privacy-policy" />
```

The main problem here is that the `hreflang` attribute is being set to the value of the custom locale, which is obviously not a valid value and is being flagged by Google as a problem in Search Console.

A secondary issue is that, since the sites are on separate domains, the different URLs are not actually alternate since they are on different sites.

I think the easiest way around this would be to be able to switch off the `hreflang="x-default"` handling via a config setting.

2. There are other instances where the custom locale value is being injected into the meta tags:

* `og:locale`: I can handle this by amending the `seomaticMeta.og` template variable to show the correct value.
* `Main Entity of Page` schema `json-ld`: The `in-language` setting uses the custom locale and doesn't appear to be editable.

Another option to solve both issues 1 and 2 might be to add a setting to map custom locales to actual locales in a config file: e.g. `en_gb_site1 => en_gb`.

**Comment by khalwat** [comment]
What version of SEOmatic are you using?

**Comment by steverowling** [comment]
1.1.54

**Comment by khalwat** [comment]
So what SEOmatic does is technically correct in both cases.

It uses whatever the main site locale is for the `x-default` so you can just re-arrange the locales to make whatever you want to be the main site locale be the first locale in the listing. Then SEOmatic will use it for the `x-default`

**Comment by steverowling** [comment]
The biggest issue is that the custom locales (`en_gb_site1`, etc. ) appear in the `hreflang` tags and these are being flagged by Search Console as incorrect language and region codes.

Of course this isn't surprising but it's the only way to get multi site working in Craft 2.

What's annoying is that actually I don't want the `hreflang` stuff at all. It's just not needed or appropriate for these sites. 

Is there a way of switching it off in SEOmatic? Or am I stuck with it if I want to use the plugin in this site?

**Comment by khalwat** [comment]
hmmmm. Maybe I can add a config setting for situations like this.

**Comment by steverowling** [comment]
That would be awesome! Seems the easiest way to handle it if you're OK with that.

**Comment by khalwat** [comment]
Try this build in the `develop` branch out:

https://github.com/nystudio107/seomatic/tree/develop

It lets you specify the `localeId`s to ignore via the `excludeLocales` in the `config.php`

**Comment by khalwat** [comment]
@steverowling How'd we do?

**Comment by steverowling** [comment]
The dev branch solves the first issues relating to `hreflang` tags, thanks!

The second issue, where the custom locale is being used for the `in-language` attribute of the `Main Entity of Page` schema `json-ld` is still there. It would be nice to fix this too as it obviously should be a valid language code in there.

Not sure how serious it is, as Google's structured data testing tool doesn't throw any errors, but the OCD part of me is bothered by it being wrong!

**Comment by khalwat** [comment]
So can you show me what it is, and what it _should be_ in your specific case @steverowling ?

**Comment by steverowling** [comment]
Sure. It's on these three sites:

* https://rtwilliams.co.uk
    custom-locale: en_gb_rtwins
    actual language: en_gb
    
```
<!-- Main Entity of Page -->

<script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebPage",
    "name": "Home",
    "description": "Commercial, Business and Personal insurance with personal advice and service from our friendly professional staff. Based in Brighton for nearly 70 years.",
    "image": {
        "@type": "ImageObject",
        "url": "https://rtwilliams.co.uk/assets/images/site/_ogTransform/RTWilliams-Insurance-OG.jpg",
        "width": "1200",
        "height": "630" 
    },
    "url": "https://rtwilliams.co.uk",
    "mainEntityOfPage": "https://rtwilliams.co.uk",
    "inLanguage": "en_gb_rtwins",
    "headline": "Home",
    "dateCreated": "2017-09-06T14:46:32+0000",
    "dateModified": "2017-10-17T13:20:32+0000",
    "datePublished": "2017-09-06T14:46:00+0000",
    "copyrightYear": "2017",
    ...
</script>
```

The `inLangauge` attribute should be `en_gb` which is a valid language code.

Of course, because I'm using a custom locale to get multi-site features in Craft 2, the custom locale value is being inserted instead.

The same is happening on the other two linked sites:

* https://rtwpropertyinsurancebrokers.co.uk/
    custom-locale: en_gb_rtwcom
    actual language: en_gb
    
* https://rtwilliamsfinancialadvisers.co.uk/
    custom-locale: en_gb_rtwifa
    actual language: en_gb
    
The custom locale is also being output by SEOmatic in the `og:locale` value, but I can handle this by amending the `seomaticMeta.og` template variable to show the correct value as this is exposed to Twig.

The `Main Entity of Page` `json-ld` isn't exposed as a template variable, so I can't use the same replacement technique I'm using for `og:locale`.

**Comment by khalwat** [comment]
So the real problem here is that we're dealing with a hack. The `int8l` settings really are intended to have the locale name _be_ the local itself.

But since you've named them diligently, what I will do is have it truncate the the locale to 4 characters, which should work pretty well.

**Comment by khalwat** [comment]
Okay @steverowling it now truncates the locales to 4 characters, so because you were smart about how you named your custom locales, it should "just work" :)

Try out the build in the `develop` branch:

https://github.com/nystudio107/seomatic/tree/develop

**Comment by steverowling** [comment]
It's close!

You need to keep the first 5 characters of the locale, so `en_gb_rtwins` becomes `en_gb`.

Currently it's returning `en_g` which isn't a valid locale.

Thanks!

**Comment by khalwat** [comment]
derp.

**Comment by khalwat** [comment]
Fixed in version 1.1.55:

https://github.com/nystudio107/seomatic/releases/tag/1.1.55
